### DISCUSSION TITLE
Is it possible to filter based on the contents of `kstack`?

### DISCUSSION THREAD

**Discussion by gsingh93** [bpftrace/bpftrace#2457] [Q&A] [answered]
This probe will print out the kernel stacktrace whenever the `kmalloc` tracepoint is called:
```
tracepoint:kmem:kmalloc
{
    print(kstack)
}
```
Is it possible to add a filter based on the stackframe, or access it's contents in an `if` statement? For example (using made-up syntax):
```
tracepoint:kmem:kmalloc
/"ksys_write" in kstack/
{
    print(kstack)
}
```
or
```
tracepoint:kmem:kmalloc
{
    if (kstack["ksys_write"]) {
        print(kstack)
    }
}
```

If this isn't currently possible, is there an issue for this, and is this something that would make sense to eventually add? If not, is this something I could do with `bcc`?

I could probably try setting a global flag in `ksys_write` first and then checking that flag in the `kmalloc` tracepoint, but this will get complex fast with multiple functions and multiple stacktraces.

**Comment by fbs** [comment, marked as answer]
The bpf call we use is `long bpf_get_stackid(void *ctx, struct bpf_map *map, u64 flags)` which only returns a single integer that identifies the unique stack. All the magic happens on the userspace side so the best we could do with this is to filter irrelevant stacks out on the userspace side. There is no option for it but you can do it with manual post processing.

There is a newer helper:  `long bpf_get_stack(void *ctx, void *buf, u32 size, u64 flags)` (https://patchwork.ozlabs.org/project/netdev/patch/20180423212752.986580-3-yhs@fb.com/) which should help to achieve what you want. But it needs quite a bit of new infrastructure to work. We need to allocate a stack buffer for the tracers and then introduce a way to iterate the buffer to see if it contains the key you want.

But yes you can do wit with bcc/libbpf!
